src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,373 @@
+                Debug.Assert(container is null || container.Id < id);
+#if DEBUG
+                _nextIdOffset = container?._nextIdOffset ?? new Random();
+#endif
+                Id = id;
+                Container = container;
+                Symbol = symbol;
+            }
+
+            internal void Free()
+            {
+                Container?.Free();
+                _variableBySlot.Free();
+                _variableTypes.Free();
+                _variableSlot.Free();
+            }
+
+            internal VariablesSnapshot CreateSnapshot()
+            {
+                return new VariablesSnapshot(
+                    Id,
+                    Container?.CreateSnapshot(),
+                    Symbol,
+                    ImmutableArray.CreateRange(_variableSlot),
+                    ImmutableDictionary.CreateRange(_variableTypes));
+            }
+
+            internal Variables CreateNestedFunction(MethodSymbol method)
+            {
+                Debug.Assert(GetVariablesForSymbol(method) is null or { Symbol: null });
+                return new Variables(id: GetNextId(), this, method);
+            }
+
+            internal int RootSlot(int slot)
+            {
+                while (true)
+                {
+                    int containingSlot = this[slot].ContainingSlot;
+                    if (containingSlot == 0)
+                    {
+                        return slot;
+                    }
+                    else
+                    {
+                        slot = containingSlot;
+                    }
+                }
+            }
+
+            internal bool TryGetValue(VariableIdentifier identifier, out int slot)
+            {
+                var variables = GetVariablesForVariable(identifier);
+                return variables.TryGetValueInternal(identifier, out slot);
+            }
+
+            private bool TryGetValueInternal(VariableIdentifier identifier, out int slot)
+            {
+                if (_variableSlot.TryGetValue(identifier, out int index))
+                {
+                    slot = ConstructSlot(Id, index);
+                    return true;
+                }
+                slot = -1;
+                return false;
+            }
+
+            internal int Add(VariableIdentifier identifier)
+            {
+                if (Count > IndexMask)
+                {
+                    return -1;
+                }
+                if (GetSlotDepth(identifier.ContainingSlot) >= MaxSlotDepth)
+                {
+                    return -1;
+                }
+                var variables = GetVariablesForVariable(identifier);
+                return variables.AddInternal(identifier);
+            }
+
+            private int AddInternal(VariableIdentifier identifier)
+            {
+                int index = Count;
+                _variableSlot.Add(identifier, index);
+                while (index >= _variableBySlot.Count)
```

Comment:
>while (index >= _variableBySlot.Count) [](start = 16, length = 38)

Is the loop really necessary? It looks like we will always do exactly one iteration. Perhaps we should simply do ```Add(identifier)```?